api/units.go

```diff
@@ -101,15 +101,33 @@ func (ur *unitsResource) set(rw http.ResponseWriter, req *http.Request, item str
 		return
 	}
 
+	newUnit := false
 	if eu == nil {
 		if len(su.Options) == 0 {
 			err := errors.New("unit does not exist and options field empty")
 			sendError(rw, http.StatusConflict, err)
+			return
 		} else if err := ValidateOptions(su.Options); err != nil {
 			sendError(rw, http.StatusBadRequest, err)
+			return
 		} else {
-			ur.create(rw, su.Name, &su)
+			newUnit = true
 		}
+	} else if eu.Name == su.Name && len(su.Options) > 0 {
+		// There is already a unit with the same name
+		// check the hashes if they do not match then we will
+		// create a new unit with the same name and later
+		// the job will be updated to this new unit.
```

Comment:
> So rather than actually replacing the existing unit in the registry, you are just adding a new one with the same name? Why do you need to do it like this? By my intuition, this is bound to cause problems...
How you think this should be done ? what problems ?

> Also, where/how will the job be updated later? And what happens to the old unit? This all needs a lot more explanation...

We instruct etcd to update the job key to point to the right target, and everything else is handled automatically...

The old unit is not referenced anymore, this may relate in part of @dongsupark PR https://github.com/coreos/fleet/pull/1510 , we may insert an explicit destroy unit hash before starting the new unit, but still not sure! IMO this should be fixed later not in this series if it really matters.